Make private Envoy sockets localhost only

meyskens opened the pull request:
This change makes parts of Envoy that do not need public access bind to
localhost only. This was previously not possible as it could not bind
on both ipv4 and v6. In Envoy 1.24 multi listeners were added making
this improvement possible.

~~Depends on https://github.com/cilium/cilium/pull/23940~~ (merged)

Fixes a part of https://github.com/cilium/cilium/issues/23353 (there are more cilium agent ports which are not managed by envoy)

```release-note
Make Envoy sockets for tproxy and the xDS API and bind to localhost only 
```

meyskens committed:
Make private Envoy sockets localhost only

This change makes parts of Envoy that do not need public access bind to
localhost only. This was previously not possible as it could not bind
on both ipv4 and v6. In Envoy 1.24 multi listeners were added making
this improvement possible.

Signed-off-by: Maartje Eyskens <maartje.eyskens@isovalent.com>

jrajahalme commented on pkg/envoy/server.go:
```diff
@@ -483,6 +483,45 @@ func getListenerAddress(port uint16, ipv4, ipv6 bool) *envoy_config_core.Address
+			SocketAddress: &envoy_config_core.SocketAddress{
+				Protocol:      envoy_config_core.SocketAddress_TCP,
+				Address:       "::1",
+				Ipv4Compat:    ipv4 && ipv6,
```
IPv4 compat likely does not work for `::1` so could leave it out.

jrajahalme commented on pkg/envoy/server.go:
```diff
@@ -483,6 +483,45 @@ func getListenerAddress(port uint16, ipv4, ipv6 bool) *envoy_config_core.Address
 	}
 }
 
+func getLocalListenerAddress(port uint16, ipv4, ipv6 bool) (*envoy_config_core.Address, []*envoy_config_listener.AdditionalAddress) {
```
Make function name plural?

jrajahalme reviewed (commented):
Nice, some early comments below.

meyskens committed:
Add IP iptables tproxy rules for local only rules

This change adds the ability to make proxy rules run over the local interface.
To do this it adds the underlying IP to the iptable ruleset for --on-ip
instead ot it defaulting to any ip.

Signed-off-by: Maartje Eyskens <maartje.eyskens@isovalent.com>

meyskens committed:
Make BPF TPROXY work with localhost sockets

This change adds a lookup to the localhost IP in the tproxy
code. It will first look for a localhost socket, failing that
it will fall back to the previous behaviour of 0.0.0.0/::

Co-authored-by: Jarno Rajahalme <jarno@isovalent.com>
Signed-off-by: Maartje Eyskens <maartje.eyskens@isovalent.com>

jrajahalme commented on bpf/lib/proxy.h:
```diff
@@ -133,11 +133,20 @@ NAME(struct __ctx_buff *ctx, CT_TUPLE_TYPE * ct_tuple, __be16 proxy_port)	\
 	tuple->SK_FIELD.dport = proxy_port;					\
-	tuple->SK_FIELD.sport = 0;						\
-	memset(&tuple->SK_FIELD.daddr, 0, sizeof(tuple->SK_FIELD.daddr));	\
+	tuple->SK_FIELD.sport = 0;					\
```
nit: unnecessary white space change before the trailing backslash.

jrajahalme commented on bpf/lib/proxy.h:
```diff
@@ -133,11 +133,20 @@ NAME(struct __ctx_buff *ctx, CT_TUPLE_TYPE * ct_tuple, __be16 proxy_port)	\
+		goto out;	\
+										\
+	/* if there's no tproxy socket on tproxy_addr look for one bound to all interfaces */ \
+	memset(&tuple->SK_FIELD.daddr, 0, sizeof(tuple->SK_FIELD.saddr));	\
```
nit: Use `sizeof(...daddr)` for consistency.

jrajahalme commented on pkg/proxy/proxy.go:
```diff
@@ -194,10 +196,12 @@ var (
+			localOnly: true,
 		},
 		DNSProxyName: {
 			proxyType: ProxyTypeDNS,
```
I would have thought DNS proxy gets the same treatment for only using local ports?

jrajahalme commented on pkg/proxy/proxy_test.go:
```diff
@@ -73,11 +74,12 @@ func (s *ProxySuite) TestPortAllocator(c *C) {
 	c.Assert(port2, Not(Equals), port)
 	c.Assert(pp.proxyType, Equals, ProxyTypeCRD)
 	c.Assert(pp.ingress, Equals, false)
+	c.Assert(pp.localOnly, Equals, true)
```
This seems weird, even though this is similar how the `ingress` parameter is handled in this case. This test would be more readable if we re-set both `ingress` and `localOnly` members when (re-)allocating the `proxyPort` after it was seen as zero in `AllocateProxyPort()`.

meyskens replied:
https://github.com/cilium/cilium/issues/24264

jrajahalme reviewed (approved):
Small nits only. Use of local only ports on DNS proxy may warrant further discussion, though.

tommyp1ckles commented on pkg/envoy/ciliumenvoyconfig.go:
```diff
@@ -361,7 +361,7 @@ func ParseResources(cecNamespace string, cecName string, anySlice []cilium_v2.XD
 	for _, listener := range resources.Listeners {
 		if listener.GetAddress() == nil {
-			port, err := portAllocator.AllocateProxyPort(listener.Name, false)
+			port, err := portAllocator.AllocateProxyPort(listener.Name, false, true)
```
Was this meant to ever be configurable?

jrajahalme replied:
It kind of is already, you can specify an address and port in the Listener spec. This path is only executed when there is no address in the Listener spec, meaning that "let each Cilium Agent pick a local port". With this change that now includes the fact the locally allocated port is listened only on localhost interfaces. The locally allocated port is essentially random, so it is different on each node and is not exposed in any user facing API. Given this it would be hard to send traffic to that port from the outside of the node anyway, so listening on localhost only does not change much. Other than have one less open port on external interfaces potentially vulnerable for port scanning.
